[AutoDiff] Support `inout` argument differentiation.

dan-zheng opened the pull request:
Includes cherry-pick of https://github.com/apple/swift/pull/29959: typing rules for `inout` parameter differentiation.

---

Add reverse-mode differentiation support for `apply` with `inout` arguments.

Notable pullback generation changes:
- If the pullback seed argument is `inout`, assign it (rather than a copy)
  directly as the adjoint buffer of the original result. This is important so
  the value is updated in-place.
- In `visitApplyInst`: skip adjoint accumulation for `inout` arguments.
  Adjoint accumulation for `inout` arguments occurs when callee pullbacks are
  applied, so no extra accumulation is necessary.

Add derivatives for functions with `inout` parameters in the stdlib for testing:
- `FloatingPoint` operations: `+=`, `-=`, `*=`, `/=`
- `Array.append`

Resolves TF-1165.

Todos:
- Add more tests, e.g. SILGen tests for `inout` derivative typing rules.
- Evaluate performance of `inout` derivatives vs functional derivatives + mutation.
- TF-1166: enable `@differentiable` attribute on `set` accessors.
- TF-1173: add forward-mode differentiation support for `apply` with `inout`
  parameters.

Exposes TF-1175: incorrect activity for class arguments.
Exposes TF-1176: incorrect activity for class `modify` accessors.
Add negative tests.

---

Examples:
```swift
// Test `inout` identity function.
func inoutIdentity(_ x: inout Float) {}
func identity(_ x: Float) -> Float {
  var result = x
  inoutIdentity(&result) // here
  return result
}
print(valueWithGradient(at: 10, in: identity))
// (value: 10.0, gradient: 1.0)
```
```swift
// Test `Float.*=` and control flow differentiation.
func product(_ array: [Float]) -> Float {
  var result: Float = 1
  for i in withoutDerivative(at: array.indices) {
    result *= array[i] // here
    // Old functional version:
    // result = result * array[i]
  }
  return result
}
print(valueWithGradient(at: [3, 4, 5], in: product))
// (value: 60.0, gradient: [20.0, 15.0, 12.0])
```
```swift
// Test `Array.append` and control flow differentiation.
func squared(_ array: [Float]) -> [Float] {
  var results: [Float] = []
  for i in withoutDerivative(at: array.indices) {
    results.append(array[i] * array[i]) // here
    // Old functional version:
    // results = results + [array[i] * array[i]]
  }
  return results
}
let v: Array<Float>.TangentVector = [4, -5, 6]
let (value, pb) = valueWithPullback(at: [1, 2, 3], in: squared)
print(value, pb(v))
// [1.0, 4.0, 9.0] [8.0, -20.0, 36.0]
```
```swift
// Test `set` accessor differentiation.
struct S: Differentiable {
  var x: Float

  var computed: Float {
    get { x }
    set { x = newValue }
  }
}
func squared(_ x: Float) -> Float {
  var s = S(x: 1)
  s.x *= x // here
  s.computed *= x // here
  return s.x
}
print(valueWithGradient(at: 10, in: squared))
// (value: 100.0, gradient: 20.0)
```

---

Thanks @marcrasi for peer debugging incorrect `inout` argument derivative values!

We discovered that `Pullback::visitApplyInst` should skip adjoint accumulation for `inout` arguments, which is interesting and not obvious. This made the first exciting test cases pass.

Working together was fun and productive 🙂

dan-zheng committed:
[AutoDiff] Type-checking support for `inout` parameter differentiation.

Semantically, an `inout` parameter is both a parameter and a result.

`@differentiable` and `@derivative` attributes now support original functions
with one "semantic result": either a formal result or an `inout` parameter.

Derivative typing rules for functions with `inout` parameters are now defined.

The differential/pullback type of a function with `inout` differentiability
parameters also has `inout` parameters. This is ideal for performance.

Differential typing rules:
- Case 1: original function has no `inout` parameters.
  - Original:     `(T0, T1, ...) -> R`
  - Differential: `(T0.Tan, T1.Tan, ...) -> R.Tan`
- Case 2: original function has a non-wrt `inout` parameter.
  - Original:     `(T0, inout T1, ...) -> Void`
  - Differential: `(T0.Tan, ...) -> T1.Tan`
- Case 3: original function has a wrt `inout` parameter.
  - Original:     `(T0, inout T1, ...) -> Void`
  - Differential: `(T0.Tan, inout T1.Tan, ...) -> Void`

Pullback typing rules:
- Case 1: original function has no `inout` parameters.
  - Original: `(T0, T1, ...) -> R`
  - Pullback: `R.Tan -> (T0.Tan, T1.Tan, ...)`
- Case 2: original function has a non-wrt `inout` parameter.
  - Original: `(T0, inout T1, ...) -> Void`
  - Pullback: `(T1.Tan) -> (T0.Tan, ...)`
- Case 3: original function has a wrt `inout` parameter.
  - Original: `(T0, inout T1, ...) -> Void`
  - Pullback: `(inout T1.Tan) -> (T0.Tan, ...)`

Resolves TF-1164.

dan-zheng committed:
[AutoDiff] Support differentiation of `apply` with `inout` arguments.

Add reverse-mode differentiation support for `apply` with `inout` arguments.

Notable pullback generation changes:
- If the pullback seed argument is `inout`, assign it (rather than a copy)
  directly as the adjoint buffer of the original result. This is important so
  the value is updated in-place.
- In `visitApplyInst`: skip adjoint accumulation for `inout` arguments.
  Adjoint accumulation for `inout` arguments occurs when callee pullbacks are
  applied, so no extra accumulation is necessary.

Add derivatives for functions with `inout` parameters in the stdlib for testing:
- `FloatingPoint` operations: `+=`, `-=`, `*=`, `/=`
- `Array.append`

Resolves TF-1165.

Todos:
- Add more tests, e.g. SILGen tests for `inout` derivative typing rules.
- Evaluate performance of `inout` derivatives vs functional derivatives + mutation.
- TF-1166: enable `@differentiable` attribute on `set` accessors.
- TF-1173: add forward-mode differentiation support for `apply` with `inout`
  parameters.

dan-zheng commented:
Note: merging `master` branch PR https://github.com/apple/swift/pull/29959 (`inout` differentiation typing rules) to `tensorflow` branch requires this PR, which has non-trivial accompanying changes to fix tests:

```
Failing Tests (5):
    Swift(linux-x86_64) :: AutoDiff/downstream/activity_analysis.swift
    Swift(linux-x86_64) :: AutoDiff/downstream/differentiation_transform_diagnostics.swift
    Swift(linux-x86_64) :: AutoDiff/downstream/differentiable_attr_type_checking.swift
    Swift(linux-x86_64) :: AutoDiff/downstream/forward_mode_diagnostics.swift
    Swift(linux-x86_64) :: AutoDiff/compiler_crashers_fixed/tf1167-differentiable-attr-override-match.swift
```

This is blocking the `master -> tensorflow` merge, so I'd like to land this PR as soon as possible.

Lesson learned: next time when doing changes that touch both upstream and downstream code, create PRs to both `tensorflow` and `master` branch, and test+merge the `tensorflow` branch PR first. This makes the `master -> tensorflow` merge process much easier.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented on test/AutoDiff/downstream/inout_parameters.swift:
```diff
@@ -0,0 +1,128 @@
+    }
+  }
+
+  // `squared` implementation implemented using a `set` accessor.
```
```suggestion
  // `squared` implemented using a `set` accessor.
```

saeta commented:
🎉

dan-zheng commented on test/AutoDiff/downstream/inout_parameters.swift:
```diff
@@ -0,0 +1,128 @@
+
+var InoutParametersTests = TestSuite("InoutParameters")
+
+// TODO: Move floating-point mutating operation tests to
```
This is tracked by TF-1173.
```suggestion
// TODO(TF-1173): Move floating-point mutating operation tests to
```

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented on stdlib/public/Differentiation/DifferentiationSupport.swift:
```diff
@@ -1236,6 +1236,25 @@ extension Array where Element : Differentiable {
 
+extension Array where Element: Differentiable {
+  @derivative(of: append)
+  public mutating func _vjpAppend(_ element: Element) -> (
```
Please make derivatives of public APIs be `@inlinable internal` since their symbols are not to be leaked.

dan-zheng replied:
Wow, thanks for the catch. Done.

rxwei commented on stdlib/public/Differentiation/DifferentiationSupport.swift:
```diff
@@ -1236,6 +1236,25 @@ extension Array where Element : Differentiable {
+  }
+
+  @derivative(of: append)
+  public mutating func _jvpAppend(_ element: Element) -> (
```
`@inlinable internal`

rxwei reviewed (commented):
Missing test cases:
1. Differentiating a function with `inout` parameters that have a `class` type. This is expected to work normally.
2. Differentiating a function whose body calls a same-file function that both takes a `Differentiable` `inout` argument and returns a `Differentiable` formal result where both the `inout` argument and the formal result are active. This is expected to trigger a differentiation error.

dan-zheng committed:
Fix SIL diff witness deserialization: result indices capacity.

Fixes `-O` test failure:

```
Failing Tests (1):
    Swift(linux-x86_64) :: AutoDiff/downstream/inout_parameters.swift
```

dan-zheng committed:
Make derivatives `@usableFromInline` + internal instead of public.

dan-zheng committed:
Add suggested tests.

rxwei commented on test/AutoDiff/downstream/inout_parameters.swift:
```diff
@@ -125,4 +126,28 @@ InoutParametersTests.test("SetAccessor") {
+  // function.
+  func inoutIdentity(_ c: inout Class) {}
+
+  func identity(_ x: Float) -> Float {
```
We need a test case where the `inout`-taking method actually *does* modify the class argument.

dan-zheng replied:
Done. I added two tests with `inout` class arguments:
- One test assigns a new value to the `inout` class argument (via `store`). This test has correct derivatives.
- The other test mutates the `inout` class argument by applying a property `modify` accessor. This test has incorrect derivatives. TF-1176 tracks the issue, I'd like to address it in a follow-up.

rxwei replied:
> * The other test mutates the `inout` class argument by applying a property `modify` accessor. This test has incorrect derivatives. [TF-1176](https://bugs.swift.org/browse/TF-1176) tracks the issue, it shouldn't be hard to fix but I'd like to address it in a follow-up.

That does not mutate the `inout` class argument. It only mutates the underlying contents, but not the argument itself. To modify the argument, you should assign something to the argument so that the class instance is changed.

dan-zheng replied:
> To modify the argument, you should assign something to the argument so that the class instance is changed.

I think that's already being done in `squaredViaMutation`. Am I missing sth?
```swift
    func squaredViaMutation(_ c: inout Class) {
      c = Class(c.x * c.x)
    }
    func squared(_ x: Float) -> Float {
      var c = Class(x)
      squaredViaMutation(&c)
      return c.x
    }
```

rxwei replied:
Oh sorry I missed this one. LGTM.

dan-zheng committed:
Add negative test for TF-1175.

Class-typed arguments are not always marked active.
This produces incorrect derivative results.

dan-zheng committed:
Revamp `inout` class argument tests.

Mutating `inout` class argument via `store`: correct derivatives.

Mutating `inout` class argument via `modify` accessor: incorrect derivatives.
Tracked at TF-1176.

rxwei commented on lib/SILOptimizer/Utils/Differentiation/JVPEmitter.cpp:
```diff
@@ -1000,10 +1000,12 @@ void JVPEmitter::prepareForDifferentialGeneration() {
   auto diffIndResults = differential.getIndirectResults();
-  assert(origIndResults.size() == diffIndResults.size());
-
+  size_t numInoutParameters = llvm::count_if(
```
Use `unsigned` instead.

dan-zheng replied:
Is there an advantage to using `unsigned` over `size_t` or `auto`? Are there established conventions?

I intentionally didn't use `unsigned` because I vaguely recall it caused a Windows-specific failure. I don't remember the exact scenario though, so the memory may be misattributed.

I actually like the idea of using `auto` whenever possible.

rxwei replied:
> Is there an advantage to using `unsigned` over `size_t` or `auto`? Are there established conventions?

We've been using `unsigned` everywhere in AutoDiff code paths. Nearly all AST/SIL APIs that return a count has `unsigned` as the result type.



> Is there an advantage to using `unsigned` over `size_t` or `auto`? Are there established conventions?
> 
> I intentionally didn't use `unsigned` because I vaguely recall it caused a Windows-specific failure. I don't remember the exact scenario though, so the memory may be misattributed.
> 
> I actually like the idea of using `auto` whenever possible.

It'd be helpful if you could point to the exact issue. Most AST/SIL/LLVM APIs are still returning `unsigned` and they work fine with Windows builds.

> I actually like the idea of using `auto` whenever possible.

I do not think that is a good idea. Some STL APIs often return `size_t` (like this one you are calling) and comparing signed and unsigned integers can lead to unexpected behavior (which I've encountered in AutoDiff code paths) that is very very hard to debug.

dan-zheng replied:
Thanks for the context, your points make sense!
Rewriting `size_t` to `unsigned` done. I'll remember to use `unsigned` in the future.

rxwei reviewed (commented):
Any forward mode tests?

dan-zheng commented:
> Any forward mode tests?

Forward-mode differentiation doesn't support `inout` arguments yet: TF-1173.
Adding support should involve straightforward changes to `JVPEmitter::visitApplyInst`.
In the meantime, forward-mode `inout` argument differentiation remains diagnosed.

dan-zheng committed:
`size_t` -> `unsigned`

Let's standardize on `unsigned`.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow
